src/MongoDB.Analyzer/Core/Builders/BuildersResolveVariables.cs

```diff
@@ -0,0 +1,445 @@
+            {
+                var expressionAnalysisContext = GenerateNewExpressionAnalysisContext(assignmentExpression, evaluationResult.SyntaxNodeToVariableValueMap);
+                variableValue = new VariableValue(expressionAnalysisContext, new List<Location>());
+            }
+
+            dirtyVariables.Add(variableName);
+            contextInformation.VariableValues[variableName] = variableValue;
+            var originalExpression = variableValue.ExpressionAnalysisContext.Node.OriginalExpression;
+
+            if (contextInformation.BuilderNodesProcessed.Add(originalExpression))
+            {
+                contextInformation.VariableValuesList.Add(variableValue);
+            }
+
+            if (assignmentExpression.Parent is ConditionalExpressionSyntax)
+            {
+                RemoveVariablesFromContext(new List<string>() { variableName }, contextInformation);
+            }
+        }
+        else
+        {
+            RemoveVariablesFromContext(new List<string>() { variableName }, contextInformation);
+        }
+
+        return evaluationResult.CanEvaluate;
+    }
+
+    private static void RemoveVariablesFromContext(IEnumerable<string> variablesToRemove, ContextInformation contextInformation)
+    {
+        var variableValues = contextInformation.VariableValues;
+        var variableValuesList = contextInformation.VariableValuesList;
+        var builderNodesProcessed = contextInformation.BuilderNodesProcessed;
+
+        foreach (var variable in variablesToRemove)
+        {
+            var isValidVariable = variableValues.TryGetValueSafe(variable, out var previousValue);
+
+            if (!isValidVariable)
+            {
+                continue;
+            }
+
+            var originalExpression = previousValue.ExpressionAnalysisContext.Node.OriginalExpression;
+
+            if (builderNodesProcessed.Add(originalExpression))
+            {
+                variableValuesList.Add(previousValue);
+            }
+        }
+
+        variableValues.RemoveRange(variablesToRemove);
+    }
+
+    private static bool TryProcessBuildersExpression(SyntaxNode syntaxNode, ContextInformation contextInformation, HashSet<string> dirtyVariables)
+    {
+        var isValidBuildersExpression = false;
+        var semanticModel = contextInformation.SemanticModel;
+
+        if (!IsValidBuilderExpression(syntaxNode))
+        {
+            RemoveVariablesFromContext(ScrapeVariables(syntaxNode), contextInformation);
+            return false;
+        }
+
+        if (syntaxNode is VariableDeclaratorSyntax variableDeclarator)
+        {
+            var rightHandAssignment = variableDeclarator.Initializer.Value as AssignmentExpressionSyntax;
+            var isDeclarationChain = rightHandAssignment != null;
+
+            isValidBuildersExpression = isDeclarationChain ? TryProcessBuildersExpression(rightHandAssignment, contextInformation, dirtyVariables) : !ContainsConditionalExpression(syntaxNode);
```

Comment:
Consider handling isDeclarationChain in single block.

Reply:
Done